### PR TITLE
Disallowing `nullable` with `AsParameters` attribute

### LINKED ISSUE #42210: Nullable `struct` not working with `AsParameter`
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Minimal APIs generate an `InvalidOperationException `when `AsParameters` attribute is applied to a `Nullable struct`.

### Expected Behavior

Since the `AsParameters` attribute will just expand all type properties as a parameter of my delegate, I expect that the `Nullable struct` works exactly the same as a struct or at least throws a better exception. (eg.: `NotSupportedException:  The nullable struct type  Nullable<T> is not supported`).

### Steps To Reproduce

``` c#
var app = WebApplication.Create(args);
app.MapGet("/products/{id}", ([AsParameters] ProductRequest? req) => Results.Ok(req.Value.Id));
app.Run();

record struct ProductRequest(int Id, int? PageSize, int? Page, ILogger<Program> Logger);
```

### Exceptions (if any)

```
System.InvalidOperationException: Body was inferred but the method does not allow inferred body parameters.
Below is the list of parameters that we found: 

Parameter           | Source                        
---------------------------------------------------------------------------------
Value               | Body (Inferred)
req                 | As Parameter (Attribute)


Did you mean to register the "Body (Inferred)" parameter(s) as a Service or apply the [FromServices] or [FromBody] attribute?


   at Microsoft.AspNetCore.Http.RequestDelegateFactory.CreateArguments(ParameterInfo[] parameters, FactoryContext factoryContext) in C:\git\aspnetcore\src\Http\Http.Extensions\src\RequestDelegateFactory.cs:line 534
   at Microsoft.AspNetCore.Http.RequestDelegateFactory.CreateTargetableRequestDelegate(MethodInfo methodInfo, Expression targetExpression, FactoryContext factoryContext, Expression`1 targetFactory) in C:\git\aspnetcore\src\Http\Http.Extensions\src\RequestDelegateFactory.cs:line 222
   at Microsoft.AspNetCore.Http.RequestDelegateFactory.Create(Delegate handler, RequestDelegateFactoryOptions options) in C:\git\aspnetcore\src\Http\Http.Extensions\src\RequestDelegateFactory.cs:line 134
   at Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions.<>c__DisplayClass26_0.<Map>g__RouteHandlerBuilderConvention|0(EndpointBuilder endpointBuilder)
   at Microsoft.AspNetCore.Routing.DefaultEndpointConventionBuilder.Build()
   at System.Linq.Enumerable.SelectListIterator`2.ToArray()
   at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.ToArray()
   at Microsoft.AspNetCore.Routing.CompositeEndpointDataSource.Initialize()
   at Microsoft.AspNetCore.Routing.DataSourceDependentCache`1.Initialize()
   at System.Threading.LazyInitializer.EnsureInitializedCore[T](T& target, Boolean& initialized, Object& syncLock, Func`1 valueFactory)
   at Microsoft.AspNetCore.Routing.Matching.DataSourceDependentMatcher..ctor(EndpointDataSource dataSource, Lifetime lifetime, Func`1 matcherBuilderFactory)
   at Microsoft.AspNetCore.Routing.Matching.DfaMatcherFactory.CreateMatcher(EndpointDataSource dataSource)
   at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.InitializeCoreAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.<Invoke>g__AwaitMatcher|8_0(EndpointRoutingMiddleware middleware, HttpContext httpContext, Task`1 matcherTask)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

### .NET Version

7.0.0 preview 5

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by brunolins16**
Fixes #42210

This PR change the `AsParameters` binding logic to not support `nullable`. That means the `RequestDelegateFactory.Create` will throw an InvalidOperationException when a nullable `class/struct` is defined with the `AsParameters` attribute, eg.:

``` c#
app.MapGet("value-type", ([AsParameters] WeatherForecastStruct? args) => Results.Ok(args)).WithOpenApi();
app.MapGet("reference-type", ([AsParameters] WeatherForecast? args) => Results.Ok(args)).WithOpenApi();

internal record WeatherForecast(DateOnly? Date, int? TemperatureC, string? Summary);
internal record struct WeatherForecastStruct(DateOnly? Date, int? TemperatureC, string? Summary);
```

The binding will be **allowed** when reference type parameter is `nullable-oblivious` (compiled before C# 8, or in a disabled context), that are nullable.

### CODE REVIEW

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 1]
```diff
@@ -4679,6 +4679,11 @@ void TestParameterListStruct([AsParameters] ParameterListStruct args)
                 args.HttpContext.Items.Add("input", args.Value);
             }
 
+            void TestParameterListNullableStruct([AsParameters] ParameterListStruct? args)
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:4682]: Add a test that doesn't pass any bindable values and check that the struct is null?

Comment (brunolins16) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:4682]: I don't know if it will ever be null based on my change.

- If any bindable values is passed, and they are required, it should cause a `BadRequestException` (or Bad Request response).

- When the properties are not required it should create an instance and have the default values for all properties.

What do you think about this?

Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:4682]: > What do you think about this?

That makes sense to me. 

> When the properties are not required it should create an instance and have the default values for all properties.

Is there a test for this?

Comment (brunolins16) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:4688]: > Is there a test for this?

Just added one.

Comment (halter73) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:4688]: Are we sure about this? This feels like the opposite direction we went with nullable `StringValues`. See the conversation at https://github.com/dotnet/aspnetcore/pull/41789#issuecomment-1181263769

We decided that it should be impossible to ever get a `default(StringValues)` as an argument to either a `StringValues` or a `StringValues?` parameter. If I declare a parameter to be nullable, I expect to be given a `null` nullable when nothing has been provided.

Are we doing the same thing for optional non-struct parameter? Are we just filling each missing property with its default value if it's marked nullable and the property is not?

What exactly happens when only a subset of required properties are provided for a nullable reference type parameter? What about a nullable struct?

Comment (brunolins16) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:4688]: I got your point about the this be a opposite direction; however, I see the `AsParameters` as an approach to create `structured parameters` where we are just organizing the original parameters in a `class`/`struct`. The actual binding is exactly the same on the properties (all default values and optionality), including the BadRequest when they are missing, and the actual `AsParameters` parameter does not contribute to the optionality (setting it as nullable will not make all properties optional).

> Are we doing the same thing for optional non-struct parameter? 

We do the same thing but as I mention, today, a nullable `AsParameters` parameter does not mean `optional`. If we can bound all properties (default value, null value or actual value from a source) an instance of the `AsParameters` parameter will be created

> What exactly happens when only a subset of required properties are provided for a nullable reference type parameter? What about a nullable struct?

The `AsParameters` parameter does not change the behavior of the binding. If the request misses a required property a Bad Request exception/response will be produced.

In general, a nullable `AsParameters` parameter should not exists :) and maybe that could be the right option, but I don't see a strong reason to throw if a nullable is defined and I prefer support it, but I would love to hear other opinions.

Comment (halter73) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:4688]: > and the actual `AsParameters` parameter does not contribute to the optionality (setting it as nullable will not make all properties optional).

Can we just throw if the `AsParameters` parameter is marked nullable since we ignore this and always pass a value anyway?

Comment (brunolins16) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:4688]: Yes we could.

>In general, a nullable AsParameters parameter should not exists :) and maybe that could be the right option, but I don't see a strong reason to throw if a nullable is defined and I prefer support it, but I would love to hear other opinions.

Comment (BrennanConroy) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:4688]: That's how I feel, nullable can be annoying to deal with in user code, plus I might expect to get a null `[AsParameters]` param value if not all values were provided by the user (which isn't what would happen). So throwing at startup would get rid of any ambiguity.

Comment (brunolins16) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:4688]: Based what I am understanding, the preference is to throw for nullable `class`/`struct` with `AsParameters`. Is that correct? If so, I will update the PR with that.

Comment (halter73) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:4688]: > Is that correct?

I think so. That's definitely my position.

Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:4688]: Yep, I agree with throwing an exception here as Bruno stated:

> In general, a nullable AsParameters parameter should not exists 

The explicit behavior of throwing an exception seems better to me here than the implicit behavior of handling setting default values for a nullable struct with AsParameters.

Comment (brunolins16) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs]: PR changed to throw when nullable.

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 2]
```diff
@@ -4,6 +4,7 @@
 #nullable enable
 
 using System.Buffers;
+using System.Data.Common;
```
Comment (brunolins16) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:7]: I will remove it :)

**Review by BrennanConroy — APPROVED** [review]
".. is not supported, mark the parameter as non-nullable."
Maybe? Idk.
